File: LeetCode/Graph/DFS/Golang/Leetcode450-delete-node-in-a-bst.go
package sbgo

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return nil
	}
	if root.Val == key {
		if root.Right == nil && root.Left == nil {
			return nil
		} else if root.Right == nil {
			return root.Left
		} else if root.Left == nil {
			return root.Right
		} else {
			ne := root.Right
			for ne.Left != nil {
				ne = ne.Left
			}
			ne.Right = deleteNode(root.Right, ne.Val)
			ne.Left = root.Left
			return ne
		}
	} else if root.Val > key {
		root.Left = deleteNode(root.Left, key)
	} else {
		root.Right = deleteNode(root.Right, key)
	}
	return root
}

File: LeetCode/Graph/DFS/JS/Leetcode450-delete-node-in-a-bst.js
var deleteNode = function (root, key) {
    if (!root) return null;
    else if (root.val == key) {
        if (!root.left && !root.right) return null;
        else if (!root.left) return root.right;
        else if (!root.right) return root.left;
        else {
            let ne = root.right;
            while (ne.left) ne = ne.left;
            ne.right = deleteNode(root.right, ne.val);
            ne.left = root.left;
            return ne;
        }
    }
    else if (root.val > key) {
        root.left = deleteNode(root.left, key);
    } else {
        root.right = deleteNode(root.right, key);
    }
    return root;
};